### ISSUE TITLE
Issue related to Batch Size

### ISSUE THREAD

**Issue by PrajwalVAtreyas** [tf-encrypted/tf-encrypted#787]
My team is using the library “tf_encrypted” for a project that we are building.
We are working towards developing a platform for data scientists to host all their models, and for clients to choose any model to use for their predictions.
We tested by running the example notebook on mnist classification.
We are trying to clone a similar cnn model for grocery classification as Raw, Ripe, or Rotten.

We have encountered an error, where the input layer is requesting for batch_size which is set to null in our model definition but is set to 32 when we call the fit function on the model being trained in normal tensorflow.keras.
The documentation specifies the batch_size is optional but raises a NotimplementedError when set to None.
Are there any updates on when layers such as LSTM, Dropout, activation functions such as softmax will be added?

It would be really grateful of you, if you could clarify and resolve our issue.

cc @mortendahl

**Comment by jvmncs** [comment]
>We have encountered an error, where the input layer is requesting for batch_size which is set to null in our model definition but is set to 32 when we call the fit function on the model being trained in normal tensorflow.keras. The documentation specifies the batch_size is optional but raises a NotimplementedError when set to None.

`batch_size` is optional because you can alternatively use `batch_input_shape`. You must provide either `batch_input_shape` or both of `batch_size` & `input_shape` when defining your Keras model in TFE, otherwise we raise an error.

Requiring the batch size when setting up a Keras model for inference was a design choice, and the explanation is somewhat technical. In order to support both [online and offline Beaver triples sources](https://github.com/tf-encrypted/tf-encrypted/blob/master/tf_encrypted/protocol/pond/triple_sources.py), any input to multiplication (or matmul) needs to have its shape defined ahead of time. This is because an offline triple source will need to know ahead of time what shape its triples should be in order for the triples it generates to be usable during the online computation (i.e. the model inference). In the future, we will likely allow None batch_size and raise an error when using this with an offline triple source, but so far we haven't had the time to add that feature.

>Are there any updates on when layers such as LSTM, Dropout, activation functions such as softmax will be added?

We definitely want to continue adding support for the full Keras API, but we also want to expand our original implementation of TFE to make it fully production-ready. This requires some additional engineering of the core codebase, which means we will have less time for adding ML features like layers for the moment. In the meantime, we're happy to accept and review pull requests adding these features. You can follow along with some of our planning [here](https://github.com/tf-encrypted/rfcs) or join our Slack if you want to chat more.

**Comment by jvmncs** [comment]
@PrajwalVAtreyas can we help you with anything else related to this issue? If not, we'll close it up

**Comment by PrajwalVAtreyas** [comment]
> @PrajwalVAtreyas can we help you with anything else related to this issue? If not, we'll close it up

Yes, thank you for clarifying the issue. We could successfully utilize tf-encrypted for our personal project and complete it because of your explanation! :) You may close it up.

**Comment by jvmncs** [comment]
That's awesome! I think we'd love to hear about your project and how your experience went. If not here, feel free to reach out in our Slack or via email `contact <at> tf-encrypted <dot> io`.

**Comment by PrajwalVAtreyas** [comment]
Yeah, i would be happy to share the outline of our project.
I am a Computer Science student at RV College of Engineering, Bangalore, India and we had to build a project for one of our course as part of the curriculum.
A team of 4, including me, @sahana63 and @haridher21, decided to make a platform for data scientists to host their models and clients to use models. Provide a front end for our pre-trained models from previous projects. We wanted to add encryption to our models to make the model serving secure. We were looking at homomorphic encryption when we came across tf-encrypted. tfe happened to have everything we were looking for. We had to remove `dropout` layers and replace `softmax` with `sigmoid` and retrain our models which were sound event classification (ANN) and grocery classification (CNN). We had to work our way around the `batch_size` problem and change the model definition in keras to make it compatible with tfe. We built the database to store users and all the models(h5, json, description and type). We wanted to add a feature where users can add more models, but due to deadline contraints we were not able to complete it. Nevertheless, our professors gave us a thumbs up :).
